Initial event-native documentation (#935)
gravitee-io/gravitee-docs: pages/apim/3.x/event-native-apim/event-native-apim-introduction.adoc

[[event-native-apim-introduction]]
= Introduction
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/event_native_apim_introduction.html
:page-folder: apim/event-native-apim
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

== Event-driven asynchronous APIs

Most developers are familiar with synchronous APIs, where the client sends a request to a server, and the server sends a response back. Every time the client needs new data, it has to send a request. You can think of synchronous APIs as a 'client pull' model. RESTful APIs are a type of synchronous API.

In contrast, asynchronous APIs are event-driven and allow you to stream data. Asynchronous APIs can use callbacks, such as webhooks, if the requested data is not available at the time of the request. You can think of asynchronous APIs as a 'server push' model. AsyncAPIs are a type of asynchronous API.

> For more information about event-driven architectures and asynchronous APIs, see the link:https://www.asyncapi.com/[AsyncAPI website].

Reviewer: AsyncApi website is about AsyncAPI specification,don't think we should rely on it to explain what is an event driven architecture and what is the benefit of it.

Author: OK, I will remove this line for now.

As this is a beta release, I don't think the documentation should describe the advantages of an event-driven architecture. Any users of this release should already know. 

The documentation for the full release should cover this.

== Event-native API management

With the rise of cloud computing, many vendors started to build their solutions for the cloud. Rather than taking on-premises software and converting it to work as SaaS, it was built from the ground to to be cloud-based. These solutions are said to be cloud-native.

> Event-native API management is similar. Rather than taking a synchronous API management solution and converting it to handle event-driven APIs, an event-native API management solution is built from the ground up to handle event-driven APIs.

Author: It is blurry, but I am using these paragraphs to link the technical documentation to the marketing material. If someone comes from the marketing material and reads this page, they will know they are looking at the right thing.

> Event-native API management is similar. Rather than taking a synchronous API management solution and converting it to handle event-driven APIs, an event-native API management solution is built from the ground up to handle event-driven APIs.

Reviewer 2: Ok got the idea !

== Gravitee's event-native API management

> Gravitee's API management solution can use an event-native policy execution mode, and can support all asynchronous, streaming, and event-driven API use cases. Of course, Gravitee's API management solution also supports synchronous use cases.

Reviewer: What do you mean by "an event-native policy execution mode" ?

Author: Sorry, that was very clumsy writing. I will change it to "Gravitee's API management solution can run in event-native mode, "

> Gravitee's API management solution can use an event-native policy execution mode, and can support all asynchronous, streaming, and event-driven API use cases. Of course, Gravitee's API management solution also supports synchronous use cases.

Author: The next section is about the new execution engine, so I don't think any link is necessary.

> Gravitee can act as a bridge between synchronous and asynchronous environments.

Reviewer: Could you elaborate?

Author: The next paragraph and the diagram elaborate a bit. This is just an introductory paragraph so I don't think we need more detail

For example, Gravitee can subscribe to a Kafka topic and then allow filtered HTTP GET subscriptions from clients as shown in the following diagram.

image:{% link /images/apim/3.x/event-native/event-native-api-management-gateway.png %}[]

> == New policy execution mode (V4 BETA)

Reviewer: ```suggestion
== New execution engine (V4 BETA)
```

> Introduced in release 3.19.0 of the Gravitee API Management platform, there is a beta version of a new policy execution engine that enables an improved execution flow for synchronous APIs and supports event-driven policy execution for asynchronous APIs. It is based on a modern and fully reactive architecture designed to address a number of challenges Gravitee users have been facing with the existing policy execution engine, available with the Gravitee's V3 API definition specification used in versions prior to APIM 3.19.x.

Reviewer: V4 engine is not only about policy execution, it changes many aspects of the execution.

== BETA Features

The following features are available in the BETA release.

=== HTTP POST and HTTP GET entrypoints

The HTTP POST and HTTP GET entrypoints are Community Edition / OSS plugins that enable external clients to publish and receive data via standard HTTP POST / HTTP GET requests, and ease the ingestion of data into event-orientated backends like Kafka.

With the HTTP POST workflow, a client application running on HTTP can post content to the Gravitee Gateway. The Gateway can then take that content and push it over to Kafka as a message that Kafka can ingest. The Gravitee Gateway acts as a protocol mediation layer in this context. This enables API owners to quickly set up a sync-to-async integration.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-http-post.png %}[]

The HTTP POST entrypoint can be used in combination with specific Async API endpoints in different ways, depending on what publish and subscribe mode actions are possible between the two when using a connector. This is demonstrated in the compatibility matrix table below:

|===

> |Entrypoint|Request message (publish mode)|Response message (subscribe mode)|Endpoint

Reviewer: You miss MQTT5.

|HTTP POST
|Supported
|Not supported
|Mock

|HTTP POST
|Supported
|Not supported
|Kafka

|HTTP POST
|Not supported
|Not supported

> |SSE

Reviewer: There is no SSE endpoint yet.

|HTTP POST
|Supported
|Not supported

> |Websocket

Reviewer: There is no websocket endpoint yet.

> |Websocket

Author: I have removed it for now. Different people seem to want different things from it, and this matrix is not complete.

This feature enables you to front your event-driven backend - currently only using Kafka via the Gravitee Community Edition / OSS Kafka connector - with Websocket or Server-Sent Events (SSE) so that API consumers can access events from the Kafka backend via one of these more consumer-friendly protocols.

For example, if an API consumer cannot connect to a native Kafka instance due to technical limitations (for example, if they cannot implement a Kafka client themselves), Kafka topics can be made available to that consumer via a more consumer-friendly protocol such as SSE or Websocket.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-event-consumption-streaming.png %}[]

This feature enables you to potentially cover use cases in the fields of Event-Driven Architecture (EDA) and Internet of Things (IoT) without the Gateway and API Management components becoming a bottleneck.

> NOTE: In comparison to the link:#http_post_and_http_get_entrypoints[HTTP POST entrypoint use case] using Kafka (where the HTTP POST method is just about sending information from the client app to Kafka), with SSE and Websocket it is possible for the client app to receive information from Kafka as well.

Reviewer: SSE is not able to publish data from client to kafka, it is only for consumption. However you can do both with websocket.

Author: Thanks. I have clarified this paragraph a bit.

This is an Enterprise Edition feature that enables you to front your backend with a link:https://en.wikipedia.org/wiki/Webhook[Webhook^].

With Webhook support, an API consumer subscribes to the Webhook API, which basically indicates to the Gateway to listen for specific Kafka messages and then call the consumer to let them know that an expected event has occurred. This communication is performed via Webhook, however it is always up to the Gateway to push the events or messages to the consumer.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-event-consumption-webhook.png %}[]

=== Security policies

The following security policies have been implemented:

* API Key: enforces API key checks during request processing (for security).
* Keyless policy: does not block any requests as it considers them as valid by default.
* JWT: generates a signed JWT with a configurable set of claims. This JWT can subsequently be forwarded to backend targets, or used in some other context.
* OAuth: checks access token validity during request processing using token introspection (for security).

> === Dual execution mode for the new execution engine

Reviewer: Don't understand what you mean here. an api V4 cannot only be executed with v4 engine, a v3 api can be executed with jupiter engine but there won't be any async support.

Author: I've removed this section.

[[event-native-apim-introduction]]
= Introduction
:page-sidebar: apim_3_x_sidebar
:page-permalink: apim/3.x/event_native_apim_introduction.html
:page-folder: apim/event-native-apim
:page-layout: apim3x

[label label-version]#New in version 3.20.0#
[label label-version]#BETA release#

== Event-driven asynchronous APIs

Most developers are familiar with synchronous APIs, where the client sends a request to a server, and the server sends a response back. Every time the client needs new data, it has to send a request. You can think of synchronous APIs as a 'client pull' model. RESTful APIs are a type of synchronous API.

> In contrast, asynchronous APIs are event-driven and allow you to stream data. Asynchronous APIs can use callbacks, such as webhooks, if the requested data is not available at the time of the request. You can think of asynchronous APIs as a 'server push' model. AsyncAPIs are a type of asynchronous API.

Reviewer 3: ```suggestion
In contrast, asynchronous APIs are event-driven and allow you to stream data. Asynchronous APIs can use callbacks, such as webhooks, if the requested data is not available at the time of the request. You can think of asynchronous APIs as a 'server push' model. AsyncAPIs are a type of asynchronous API.

In contrast, asynchronous APIs are event-driven, which means that requests and responses are sent when certain events occur that serve as triggers. So, for example, you can send a request triggered by a specific event type, and you would then get the response for that request at a later point in time, as and when it becomes available the respective response-triggering event occurs. Asynchronous APIs also allow you to stream data - a common use case in banking and finance systems and applications, as well as many other domains.
```

Event-native API management is similar. Rather than taking a synchronous API management solution and converting it to handle event-driven APIs, an event-native API management solution is built from the ground up to handle event-driven APIs.

== Gravitee's event-native API management

Gravitee's API management solution can run in event-native mode, and it can support all asynchronous, streaming, and event-driven API use cases. Of course, Gravitee's API management solution also supports synchronous use cases.

Gravitee can act as a bridge between synchronous and asynchronous environments.

For example, Gravitee can subscribe to a Kafka topic and then allow filtered HTTP GET subscriptions from clients as shown in the following diagram.

image:{% link /images/apim/3.x/event-native/event-native-api-management-gateway.png %}[]

== BETA Features

The following features are available in the BETA release.

=== HTTP POST and HTTP GET entrypoints

> The HTTP POST and HTTP GET entrypoints are Community Edition / OSS plugins that enable external clients to publish and receive data via standard HTTP POST / HTTP GET requests, and ease the ingestion of data into event-orientated backends like Kafka.

Reviewer 3: ```suggestion
The HTTP POST and HTTP GET entrypoints are Community Edition / OSS plugins that enable external clients to publish and receive data via standard HTTP POST / HTTP GET requests, and ease the ingestion of data into event-oriented backends like Kafka.
```

Reviewer 3: NB We use American English for our docs therefore the change.

With the HTTP POST workflow, a client application running on HTTP can post content to the Gravitee Gateway. The Gateway can then take that content and push it over to Kafka as a message that Kafka can ingest. The Gravitee Gateway acts as a protocol mediation layer in this context. This enables API owners to quickly set up a sync-to-async integration.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-http-post.png %}[]

> The HTTP POST entrypoint can be used in combination with specific Async API endpoints in different ways, depending on what publish and subscribe mode actions are possible between the two when using a connector. 

Reviewer 2: yep, the HTTP POST offer the ability to push messages to any supported event broker via a "sync call" (post)

> The HTTP POST entrypoint can be used in combination with specific Async API endpoints in different ways, depending on what publish and subscribe mode actions are possible between the two when using a connector. 

Reviewer 3: ```suggestion
The HTTP POST entrypoint enables you to push messages to any supported event broker via a "sync call" (post) - it can be used in combination with specific Asynchronous API endpoints in different ways, depending on what publish and subscribe mode actions are possible between the two when using a connector. 
```

=== Support for Websocket and SSE entrypoints

Support for Server-Sent Events (Enterprise Edition only) and Websocket (Community Edition/OSS) entrypoints are provided as separate plugins, available from the link:https://download.gravitee.io/#graviteeio-apim/plugins/entrypoints/[entrypoints plugin download page]

> This feature enables you to front your event-driven backend - currently only using Kafka via the Gravitee Community Edition / OSS Kafka connector - with Websocket or Server-Sent Events (SSE) so that API consumers can access events from the Kafka backend via one of these more consumer-friendly protocols.

Reviewer 2: Kafka & MQTT5

Author: Sorry, I don't understand this comment. What change is needed?

Reviewer 2: it says "currently only using Kafka via OSS Kafka connector" but we got an EE version + MQTT5 connector

> This feature enables you to front your event-driven backend - currently only using Kafka via the Gravitee Community Edition / OSS Kafka connector - with Websocket or Server-Sent Events (SSE) so that API consumers can access events from the Kafka backend via one of these more consumer-friendly protocols.

Author: ```suggestion
This feature enables you to front your event-driven backend - currently only using Kafka or the MQTT5 connectors  - with Websocket or Server-Sent Events (SSE) so that API consumers can access events from the Kafka backend via one of these more consumer-friendly protocols.
```

NOTE: In comparison to the link:#http_post_and_http_get_entrypoints[HTTP POST entrypoint use case] using Kafka (where the HTTP POST method is just about sending information from the client app to Kafka), with Websocket it is possible for the client app to send and receive information from Kafka as well. SSE can consume information from Kafka.

=== Webhook entrypoint

This is a feature that enables you to front your backend with a link:https://en.wikipedia.org/wiki/Webhook[Webhook^]. It has CE and EE versions.

With Webhook support, an API consumer subscribes to the Webhook API, which basically indicates to the Gateway to listen for specific Kafka messages and then call the consumer to let them know that an expected event has occurred. This communication is performed via Webhook, however it is always up to the Gateway to push the events or messages to the consumer.

The diagram below shows the workflow for this use case:

image:{% link /images/apim/3.x/event-native/event-native-api-management-use-case-event-consumption-webhook.png %}[]

=== Security policies

> The following security policies have been implemented:

Reviewer 2: Should we mention that we keep the same behavior as previously on this ?

Author: OK, how does "As with the previous version, the following security policies have been implemented" sound?

Reviewer 2: sounds good

> The following security policies have been implemented:

Author: ```suggestion
As with the previous version, the following security policies have been implemented:
```
